### ISSUE TITLE
Data preprocessing mechanism in SQLFLow.

### ISSUE THREAD

**Issue by Echo9573** [sql-machine-learning/sqlflow#1264]
Data preprocessing usually includes not only the general statistical classes but also the following processing (see the figure below). Some data processing can be easily implemented using SQL, while others are more difficult. It is important to embed the data preprocessing mechanism into SQLFLow.
The following is one of the ways of data preprocessing.
![image](https://user-images.githubusercontent.com/24820352/69639757-10bb3880-1098-11ea-9a16-3156d301a034.png)
1. Load data preprocessing: The original data is divided into a training set and test set. The index col and target col of the training set (test set) are reserved columns, and the remaining columns are used as feature columns (x_trainset) for data preprocessing.
2. Missing and outlier detection:
    1. Delete rows (remove duplicate data).
    1. Delete columns (data missing exceeds threshold a, numeric values ​​are set to null if field value exceeds threshold range b).
3. Data type identification (mostly can be implemented by feature column), there are some processing details such as: in the category field, if a certain value is less than a certain threshold (such as 5%), it is classified into other categories.
4. Null value filling:
    1. For numeric variables, lai is usually filled with null values.
    1. Numeric Null Value Filling: Grouped by category, where the category is a certain feature combination set, such as the combination of the two characteristics of "gender" and "car brand", and the null value is randomly selected in the corresponding set, for example (gender I am a male, and I randomly select a value from the feature set of Volkswagen).
5. Feature scaling and one-hot encoding of categorical data

**Note:**
The test set needs to be processed according to the preprocessing of the training set. For example, operations such as deleting columns need to be aligned with the training set.

**Comment by tonyyang-svail** [comment]
Hi @Echo9573, thanks for submitting this comprehensive list. Could you please mention which item can be implemented by SQL and which item can't?

**Comment by Echo9573** [comment]
> Hi @Echo9573, thanks for submitting this comprehensive list. Could you please mention which item can be implemented by SQL and which item can't?

Hi, All these operations are possible using SQL for processing. It should be noted that this is a fit-transform process, which records the pre-processing operations performed on the training set, and then uses the same rules to operate the test set / validation set .

**Comment by typhoonzero** [comment]
Just for recording, for null value filling, a sample SQL can be like:

```sql
select sex,embarked, age, Coalesce(age, 
    (select b.age from samples b 
    where a.sex=b.sex and a.embarked=b.embarked and b.age is not NULL order by RAND() limit 1))
from samples a;
```

**Comment by xieliaing** [comment]
can we use SQL view or SQL procedure for this ? We can even publish some templates.
I am afraid these new add-ons will put too much burden on SQLFlow and make SQLFlow deviate from its original purpose.

**Comment by typhoonzero** [comment]
Totally agree. Currently most common use of doing pre-processing using SQL is like `create table as SELECT ...` to create temporary tables and process the data step by step.

**Comment by weiguoz** [comment]
> Totally agree. Currently most common use of doing pre-processing using SQL is like `create table as SELECT ...` to create temporary tables and process the data step by step.

I have tried using `SQL view` in Maxcompute for splitting the input dataset(into a training `view` and a validation `view`), but it failed for ALPS to consume a `view` dataset.
I'm not sure if it will succeed in the hive+TF(or XGBoost).
